jl4-core/src/L4/EvaluateLazy.hs

```diff
@@ -0,0 +1,849 @@
+evalTopDecl _env (Import _ann _import_) =
+  pure []
+
+evalDirective :: Environment -> Directive Resolved -> Eval [EvalDirective]
+evalDirective env (LazyEval _ann expr) =
+  pure ((\ r -> MkEvalDirective r expr env) <$> toList (rangeOf expr))
+evalDirective _env (Eval _ann _expr) =
+  pure []
+evalDirective _env (Check _ann _expr) =
+  pure []
+
+evalLocalDecl :: Environment -> LocalDecl Resolved -> Eval ()
+evalLocalDecl env (LocalDecide _ann decide) =
+  evalDecide env decide
+evalLocalDecl env (LocalAssume _ann assume) =
+  evalAssume env assume
+
+-- We are assuming that the environment already contains an entry with an address for us.
+evalAssume :: Environment -> Assume Resolved -> Eval ()
+evalAssume env (MkAssume _ann _tysig (MkAppForm _ n []   _maka) _) =
+  updateTerm env n (WHNF (ValAssumed n))
+evalAssume env (MkAssume _ann _tysig (MkAppForm _ n _args _maka) _) = do
+  -- TODO: we should create a given here yielding an assumed, but we currently cannot do that easily,
+  -- because we do not have Assumed as an expression, and we also cannot embed values into expressions.
+  updateTerm env n (WHNF (ValAssumed n))
+
+updateTerm :: Environment -> Resolved -> Thunk -> Eval ()
+updateTerm env n thunk =
+  case lookupTerm env n of
+    Nothing -> exception (RuntimeScopeError n)
+    Just rf -> updateThunk rf thunk
+
+-- We are assuming that the environment already contains an entry with an address for us.
+evalDecide :: Environment -> Decide Resolved -> Eval ()
+evalDecide env (MkDecide _ann _tysig (MkAppForm _ n []   _maka) expr) =
+  updateTerm env n (Unevaluated expr env)
+evalDecide env (MkDecide _ann _tysig (MkAppForm _ n args _maka) expr) = do
+  let
+    v = ValClosure (MkGivenSig emptyAnno ((\ r -> MkOptionallyTypedName emptyAnno r Nothing) <$> args)) expr env
+  updateTerm env n (WHNF v)
+
+-- We are assuming that the environment already contains an entry with an address for us.
+evalDeclare :: Environment -> Declare Resolved -> Eval ()
+evalDeclare env (MkDeclare _ann _tysig _appFormAka t) =
+  evalTypeDecl env t
+
+evalTypeDecl :: Environment -> TypeDecl Resolved -> Eval ()
+evalTypeDecl env (EnumDecl _ann conDecls) =
+  traverse_ (evalConDecl env) conDecls
+evalTypeDecl env (RecordDecl _ann mcon tns) =
+  traverse_ (\ c -> evalConDecl env (MkConDecl emptyAnno c tns)) mcon
+evalTypeDecl _env (SynonymDecl _ann _t) =
+  pure ()
+
+evalConDecl :: Environment -> ConDecl Resolved -> Eval ()
+evalConDecl env (MkConDecl _ann n []) =
+  updateTerm env n (WHNF (ValConstructor n []))
+evalConDecl env (MkConDecl _ann n tns) = do
+  -- constructor
+  updateTerm env n (WHNF (ValUnappliedConstructor n))
+  conRef <- ref (TypeCheck.getName n) n
+  -- selectors (we need to create fresh names for the lambda abstractions so that every binder is unique)
+  traverse_ (\ (i, MkTypedName _ sn _t) -> do
+    arg    <- def (TypeCheck.getName n)
+    argRef <- ref (TypeCheck.getName n) arg
+    args <- traverse def (TypeCheck.getName <$> tns)
```

Comment:
is there a specific reason why we cannot do this traversal on the outside of the `travesrse_` on line 726 and then instead of doing the indexing, just zip them with `ts`?

Reply:
every selector is its own lambda.